Docs extproc (#9073)
solo-io/gloo: docs/content/guides/traffic_management/extproc/about.md

---
title: About external processing
weight: 40
description: Learn about what external processing is, how it works, and how to enable it in Gloo Edge. 
---

> Envoy offers multiple filters that you can use to manage, monitor, and secure traffic to your apps. However, although Envoy is extensible via C++ and WebAssembly modules, it might not be practical to implement these extensions for all of your apps. You might also have very specific requirements for how to process a request or response to allow traffic routing between different types of apps, such as adding specific headers to new and legacy apps. 

Reviewer: ```suggestion
Envoy offers multiple filters that you can use to manage, monitor, and secure traffic to your apps. Although Envoy is extensible via C++ and WebAssembly modules, it might not be practical to implement these extensions for all of your apps. You might also have very specific requirements for how to process a request or response to allow traffic routing between different types of apps, such as adding specific headers to new and legacy apps. 
```

> With external processing, you can implement an external processing server that can read and modify all aspects of an HTTP request or response, such as headers, body, and trailers giving you the flexibility to apply your requirements to all types of apps without the need to run WebAssembly or other custom scripts.  

Reviewer: ```suggestion
With external processing, you can implement an external processing server that can read and modify all aspects of an HTTP request or response, such as headers, body, and trailers. This gives you the flexibility to apply your requirements to all types of apps, without the need to run WebAssembly or other custom scripts. 
```

{{% notice note %}}
External processing is an Enterprise-only feature. 
{{% /notice %}}

{{% notice warning %}}

> Envoy's external processing filter is considered a work in progress and has an unknown security posture. Do not use this feature in production environments. 

Reviewer: is there a link to the envoy docs for an official feature status?

> Envoy's external processing filter is considered a work in progress and has an unknown security posture. Do not use this feature in production environments. For more information, see the [Envoy documentation](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ext_proc_filter#external-processing).

Reviewer 2: ```suggestion
Envoy's external processing filter is considered a work in progress and has an unknown security posture. Use caution when using this feature in production environments. For more information, see the [Envoy documentation](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ext_proc_filter#external-processing).
```
i don't know if we should outright say not to use it in production; they should use it with caution

The extProc server is a gRPC interface that must be able to respond to events in the lifecycle of an HTTP request. When the extProc filter is enabled in Gloo Edge and a request or response is received on the gateway, the filter communicates with the extProc server by using bidirectional gRPC streams.

To implement your own extProc server, make sure that you follow [Envoy's technical specification for an external processor](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_proc/v3/ext_proc.proto#extensions-filters-http-ext-proc-v3-externalprocessor). You can also follow the [Header manipulation example]({{% versioned_link_path fromRoot="/guides/traffic_management/extproc/header-manipulation/" %}}) to try out extProc in Gloo Edge with a sample extProc server.

## Enable extProc in Gloo Edge

You can enable extProc for all requests and responses that the gateway processes by using the [Settings]({{% versioned_link_path fromRoot="/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/settings.proto.sk/" %}}) custom resource. Alternatively, you can enable extProc for a specific gateway listener, virtual host, or route. 

> Example Settings configuration: 

Reviewer: would this need to go in the `default` settings cr?

Reviewer: okay sounds like yes, saw this in the next file - i would just clarify that
```suggestion
Example configuration to add to the `default` Settings resource: 
```
